### ISSUE TITLE
Provide an World::query method, which does not borrow a mutable instance of world.

### ISSUE THREAD

**Issue by Netzeband** [bevyengine/bevy#11339]
## What problem does this solve or what need does it fill?

In a unit test of systems, it is often required to do several queries to check the attributes of components, after systems run on them.
In general those checks only require immutable references to those components, however the method `World::query(&mut self)` takes a mutable instance of the world, regardless of the type of query. Therefore there can only be one reference of the query result at a time. As soon, as one needs a second reference of this or another query, with overlapping lifetimes, it will not work anymore, because then multiple mutable borrows of `World` are done.

A nice example how this prevents testing, is an example with a parent component `ComponentA`, which has child components `ComponentB`. If inside a test, you want to check, that the children of a certain parent have specific conditions, then you would do it like this: 

```rust
    use bevy::prelude::*;

    #[derive(Component)]
    struct ComponentA;

    #[derive(Component)]
    struct ComponentB;

    #[test]
    fn test_that_does_not_work() {
        let mut app = App::new();

        let parent_entity = app.world.spawn(ComponentA).with_children(|builder| {
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
        }).id();

        // here run a system under test ...

        let children = app.world.query_filtered::<&Children, With<ComponentA>>()
            .get(&app.world, parent_entity)
            .unwrap();

        for &child_entity in children {
            if let Ok(component_b) = app.world.query::<&ComponentB>().get(&app.world, child_entity) {
                // check something here ...
            }    
        }
    }
```

This does not work, since the immutable reference `children` (`&Children`) is still borrowing `app.world` mutable at the same time as I want to query the `&ComponentB` inside the loop, which also requires references of `app.world`. 

## What solution would you like?

I would like to have two distinct methods: `World::query` for queries, which only returns immutable references and `World::query_mut`, which can return mutable references, but then is blocking any further query for that time. 

## What alternative(s) have you considered?

If course in the example above it would be possible to first collect all children entities in a vector before using them to query the reference to `&ComponentB` in the loop:

```rust
    use bevy::prelude::*;

    #[derive(Component)]
    struct ComponentA;

    #[derive(Component)]
    struct ComponentB;

    #[test]
    fn test_that_works() {
        let mut app = App::new();

        let parent_entity = app.world.spawn(ComponentA).with_children(|builder| {
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
        }).id();

        // here run a system under test ...

        let children = app.world.query_filtered::<&Children, With<ComponentA>>()
            .get(&app.world, parent_entity)
            .unwrap();

        let children_list: Vec<Entity> = children.iter().map(|x| *x).collect();

        for child_entity in children_list {
            if let Ok(component_b) = app.world.query::<&ComponentB>().get(&app.world, child_entity) {
                // check something here ...
            }    
        }        
    }
``` 

Since we copy the `Entity` objects into the vector, the mutable reference to `World` is not necessary anymore, when we loop over them.
However, keep in mind, that this is just a minimal example. In real scenarios, it is not always possible to copy data from one query into local variables to use them later in the other query. For example, if you need two immutable references from two different components at the same time to perform the check.

I found myself writing a lot of code, which is just storing a ridiculous amount of data in local vectors, just to workaround the issue above. 

Another possibility would be to write a system, which performs the checks. Because in systems, I can have as many queries and references from those queries as I wish. However, it introduces another form of indirection to the tests, so that they are harder to read.

## Additional context

Nothing

**Comment by SkiFire13** [comment]
> In general those checks only require immutable references to those components, however the method `World::query(&mut self)` takes a mutable instance of the world, regardless of the type of query. Therefore there can only be one reference of the query result at a time.

The result of `World::query` does not borrow the `World`, it only temporarily needs the world to e.g. register components that might be missing. This creates a `QueryState`, which is just the state needed by a `Query` and doesn't borrow anything from the `World`, instead the borrowing happens in the call to `get`/`get_mut`/`iter`/`iter_mut`/etc etc (which as you can see already have different APIs for `ReadOnlyWorldQuery`s and not).

The problem in your example is that you're keeping the result of `get` while you call the second `query`, but there's no reason that second call to `query` has to be there. In fact you can create all the `QueryState`s at the start, and then call `get`/etc etc when needed.

```rs
    use bevy::prelude::*;

    #[derive(Component)]
    struct ComponentA;

    #[derive(Component)]
    struct ComponentB;

    #[test]
    fn test_that_does_not_work() {
        let mut app = App::new();

        let mut children_state = app.world.query_filtered::<&Children, With<ComponentA>>();
        let mut component_b_state = app.world.query::<&ComponentB>();

        let parent_entity = app.world.spawn(ComponentA).with_children(|builder| {
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
        }).id();

        // here run a system under test ...

        let children = children_state.get(&app.world, parent_entity).unwrap();

        for &child_entity in children {
            if let Ok(component_b) = component_b_state.get(&app.world, child_entity) {
                // check something here ...
            }    
        }
    }
```

Another similar option, which allows you to more closely mimic a system, is to use `SystemState`. Its generic argument is a tuple of `SystemParam`s, like those you would have in a system. Its `new` is similar to `.query()` in the sense that it requires a `&mut World` but the result doesn't borrow from it, so you're better off using it when you aren't borrowing from the `World` (e.g. at the start), and its `.get(&world)`/`.get_mut(&mut world)` are the ones borrowing the world, returning the same parameters your system would receive. This way you can write your code in the same you write a system, without however needing an additional function for it. This way you can also have multiple `Query`s that require mutable access to components (of course not the same components, like it normally works in systems), which you would normally not be able to have access to with only `.query()`.

```rs
    use bevy::prelude::*;

    #[derive(Component)]
    struct ComponentA;

    #[derive(Component)]
    struct ComponentB;

    #[test]
    fn test_that_does_not_work() {
        let mut app = App::new();

        let mut system_state = SystemState::<(
            Query<&Children, With<ComponentA>>,
            Query<&ComponentB>,
        )>::new(&mut app.world);

        let parent_entity = app.world.spawn(ComponentA).with_children(|builder| {
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
            builder.spawn(ComponentB);
        }).id();

        // here run a system under test ...
        let (children_query, component_b_query) = system_state.get(&app.world);

        let children = children_query.get(parent_entity).unwrap();

        for &child_entity in children {
            if let Ok(component_b) = component_b_query.get(child_entity) {
                // check something here ...
            }    
        }
    }
```

**Comment by Netzeband** [comment]
Wow! Thank you very much for the fast and comprehensive answer. Then I completely misunderstood it so far. 

You are right, I tried out to instantiate the query states in the beginning and use them later and it works very well. I also gained a lot of new understanding how bevy is built internally. And your second solution is perfect for more complex situations. 

Maybe someone can change the labels of this issue to be a Question. I guess its answer will be very helpful for many people.
